### PR TITLE
"added user sign up, convert/update guest user to registered user during signup with update existing guest user created in DB"

### PR DESCRIPTION
**Pull request by amolsaxe3**
1) "added user sign up, it converts existing guest user to registered user during signup by making update user for that specific guest user id"
2) **Known bug**: after redirect using props.history.push(`/`) still doesnt reflect logged in register user. On hard page refresh it does takes the registered user in the nav bar. No error shown in console.
3) Database updates to existing user is fine.
4) need auth review/any specific check if may be missing for that bug fix or otherwise.

### CODE REVIEW

**Comment by amolsaxe3** [comment]
5) signup auth bug - behavior noted below. User id started with guest then same converted to registered user, same session id. some auth issue requiring hard refresh. also same creds re-login is failing/auth check bug potentially during sign up.

File: app/components/login/LoginPage.js [hunk 1]
```diff
@@ -83,7 +83,11 @@ function LoginPage(props) {
           className={classes.formElem}
         />
         <Button className={classes.formElem} variant="contained" color="primary" onClick={handleSubmit}>Log In</Button>
+              <a href="#/signup" variant="body2">
```
Comment (ajmorozoff) [inline app/components/login/LoginPage.js]: use router-dom Link component instead

File: app/components/login/NewUserRegistration.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        <TextField
+          id="password"
+          label="Password"
+          variant="outlined"
+          type="password"
+          value={password}
+          onChange={handleChange}
+          className={classes.formElem}
+        />
+        <Button className={classes.formElem} variant="contained" color="primary" onClick={handleSubmit}>Log In</Button>
```
Comment (ajmorozoff) [inline app/components/login/NewUserRegistration.js]: The copy here should be 'Create Account' not 'Log In'
Be careful if you're just copying/pasting huge blocks of code.

File: app/components/login/NewUserRegistration.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        user.isRegistered &&
+        <div>
+          LoggedIn!
+                    </div>
+      }
+      <form className={classes.root}>
+        <h1 className={classes.header}>Sign up</h1>
+        <h3 className={classes.header}>glad to see you!</h3>
+        {loginError &&
+          <ErrorBar title={"Login failed"} message={"Incorrect email or password"} />
```
Comment (ajmorozoff) [inline app/components/login/NewUserRegistration.js]: why is the login error still here with incorrect email/password messages? This is new user creation.
Again, don't just copy/paste huge code blocks.

File: app/components/login/NewUserRegistration.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    setInit(false);
+    setError(false);
+    dispatch(updateUser(user.id, { firstName, lastName, email, password, userId: user.id }));
+    setEmail('');
+    setPassword('');
+    props.history.push(`/`)
+  }
+
+  useEffect(() => {
+    if (user && user.isRegistered) {
```
Comment (ajmorozoff) [inline app/components/login/NewUserRegistration.js]: this logic doesn't make sense -- a user shouldn't be able to get to this page via the UI if they're logged in. I would just redirect if user.isRegistered

File: app/components/login/NewUserRegistration.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      props.history.push('/');
+    }
+    if (!initLoad && !user.isRegistered) {
+      setError(true);
+    }
+    if (email || password) {
+      setError(false);
```
Comment (ajmorozoff) [inline app/components/login/NewUserRegistration.js]: Again, all this error logic was for the login and needs to be modified for account creation if you're going to surface errors

File: app/components/login/NewUserRegistration.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        break;
+      default:
+        break;
+    }
+  };
+
+  const handleSubmit = () => {
+    setInit(false);
+    setError(false);
+    dispatch(updateUser(user.id, { firstName, lastName, email, password, userId: user.id }));
```
Comment (ajmorozoff) [inline app/components/login/NewUserRegistration.js]: I recommend adding a check to see if the user email already exists in the database first

Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: The guest user is converted to actual user and email field will always be filled with dummy email per guest user creation flow, unless we keep it empty?? 
So check is not needed (?)

Comment (ajmorozoff) [reply app/components/login/NewUserRegistration.js]: Part of converting a guest to a real user should be asking the user to provide their real email. 
We should check if their real email already exists in the database, and return an error "you already have an account with this email"
If the email isn't there, then update the user record in the database with the real email.

File: app/components/login/LoginPage.js (hunk 1 above)
Comment (amolsaxe3) [reply app/components/login/LoginPage.js]: Done.

File: app/components/login/NewUserRegistration.js (hunk 2 above)
Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: Done. :) oversight error at 3 am coding side effect.

File: app/components/login/NewUserRegistration.js (hunk 3 above)
Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: Done.

File: app/components/login/NewUserRegistration.js (hunk 6 above)
Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: This only user creation screen and account management/edit should be separate functionality, not part of signup page ideally.

Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: if valid email exists check will have to be done. Will try if can be done tonight else will be done later.

File: app/components/login/NewUserRegistration.js (hunk 4 above)
Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: Done

File: app/components/login/NewUserRegistration.js (hunk 5 above)
Comment (amolsaxe3) [reply app/components/login/NewUserRegistration.js]: Done

**Comment by lachlan89** [comment]
FYI there are merge conflicts here

**Pull request closed without merging**